### ISSUE TITLE
Roslyn doesn't correctly update nullability on += operators for events

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#42557]
**Version Used**: 
```
Branch master (10 Mar 2020)
Latest commit 441c154 by msftbot[bot]:
Merge pull request #42303 from CyrusNajmabadi/useCompoundParens

Support parenthesized expressions in use-compound-operator.
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
public delegate void D();
class C
{
  public event D? ne;
  public D? df;
  void M()
  {
    df += delegate { };
    df(); // no warnings
      
    ne += delegate { };
    ne(); // false positive CS8602 derefernece of a possible null reference
  }
}
```

**Expected Behavior**:
No warnings. The event cannot be null

**Actual Behavior**:
`CS8602 Dereference of a possible null reference` is reported for the event usages

**Comment by YairHalberstadt** [comment]
Given that events are often added/removed on a different thread to where they are run, maybe this is for the best, since you can  prove another thread didn't remove it?

**Comment by TessenR** [comment]
I believe Roslyn doesn't care about different threads.
But `+=` and `=` should be handled the same way anyway as the same argument can be made about regular assignments and Roslyn doesn't warn you about them even if you remove the event handler in the same method. E.g. the following code crashes with a null reference exception but doesn't have any warnings.

```cs
#nullable enable
using System;
public class C {
    public static event EventHandler? e;
    public static void Main() {
        EventHandler del = delegate { };
        e = del;
        e -= del;
        e(null, null!);
    }
}
```

**Comment by YairHalberstadt** [comment]
@TessenR 

In general nullability analysis doesn't care about different threads, but given that events are designed so that you are meant add or remove them on a different thread, it might make sense to make an exception here.

**Comment by TessenR** [comment]
You argument about different threads would mean that events should never have a trackable nullable state and always use the state declared by the event itself. This is not how it works now for any other form of null check e.g. regular assignments and regular null checks. I do not see why you think making an exception only for `+=` would make it any better. 

Considering `e` not nullable after `e = (EventHandler) Delegate.Combine(e, del);` but not `e += del` is not consistent as they are the same.

The following samples should both have warnings under assumption that different thread can change the event at any time. However only the first one reports a warning. I think this is not correct. Of course Roslyn team can explain why I'm wrong or decide that they never intended to track events and add warnings to the second sample.
```
e += del;
// here the delegate might be removed from another thread and event will become null
e();
```
vs
```
if (e != null)
{
  // here the delegate might be removed from another thread and event will become null
  e();
}
```

**Comment by YairHalberstadt** [comment]
```csharp
if (e != null)
{
  // here the delegate might be removed from another thread and event will become null
  e();
}
```

This has been a famous example for years of what not to do. I would argue the correct thing to do would be to have an analyser that warns in that case to use `?.Invoke()` instead.

**Comment by TessenR** [comment]
The problem is that either all of these code samples should have a warning or neither of them. 

I didn't say this is a good pattern. I didn't say it shouldn't report warnings. I said that it currently does not report warnings. It does so because Roslyn team decided to ignore the possibility of member access from another thread for all members including events. Under this assumption (no other threads exist) `+=` is processed incorrectly. I only want the compiler to report consistent diagnostics under selected set of assumptions. 

Please feel free to create an issue about reporting a warning in this code as it currently doesn't report it. If you will persuade the team to change their assumptions to not track local state for events I would argue that all of the examples above should have warnings. But I will still all of them should have the same set of warnings.

**Comment by jcouv** [comment]
Relates to https://github.com/dotnet/roslyn/issues/29839 (Are nullable annotations on events meaningful/allowed?)

**Comment by RikkiGibson** [comment]
I want to address this issue as it is a point of incongruity with the old vs. new nullable constructor analysis. The event needs to update to a not-null state after adding an event accessor to it, otherwise we give a warning on exit that the event is required to be not-null it is maybe-null.
